discovery/openstack/host.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2017 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package openstack
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/hypervisors"
+	"github.com/gophercloud/gophercloud/pagination"
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/common/model"
+	"golang.org/x/net/context"
+
+	"github.com/prometheus/prometheus/config"
+)
+
+const (
+	openstackLabelHostIP         = openstackLabelPrefix + "host_ip"
+	openstackLabelHostName       = openstackLabelPrefix + "host_name"
+	openstackLabelHostStatus     = openstackLabelPrefix + "host_status"
+	openstackLabelHostState      = openstackLabelPrefix + "host_state"
+	openstackLabelHostHypervisor = openstackLabelPrefix + "host_hypervisor_type"
+)
+
+// HostDiscovery discovers OpenStack hosts.
+type HostDiscovery struct {
+	authOpts *gophercloud.AuthOptions
+	region   string
+	interval time.Duration
+	port     int
+}
+
+// NewHostDiscovery returns a new host discovery.
+func NewHostDiscovery(opts *gophercloud.AuthOptions,
+	interval time.Duration, port int, region string) *HostDiscovery {
+	return &HostDiscovery{authOpts: opts, region: region, interval: interval, port: port}
+}
+
+// Run implements the TargetProvider interface.
+func (h *HostDiscovery) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	// Get an initial set right away.
+	tg, err := h.refresh()
+	if err != nil {
+		log.Error(err)
+	} else {
+		select {
+		case ch <- []*config.TargetGroup{tg}:
+		case <-ctx.Done():
+			return
+		}
+	}
+
+	ticker := time.NewTicker(h.interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			tg, err := h.refresh()
+			if err != nil {
+				log.Error(err)
+				continue
+			}
+
+			select {
+			case ch <- []*config.TargetGroup{tg}:
+			case <-ctx.Done():
+				return
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (h *HostDiscovery) refresh() (tg *config.TargetGroup, err error) {
+	t0 := time.Now()
+	defer func() {
+		refreshDuration.Observe(time.Since(t0).Seconds())
+		if err != nil {
+			refreshFailuresCount.Inc()
+		}
+	}()
+
+	provider, err := openstack.AuthenticatedClient(*h.authOpts)
+	if err != nil {
+		_ = fmt.Errorf("could not create OpenStack session: %s", err)
```

Comment:
This line doesn't seem to do anything. Is this on purpose?

Reply:
Updated.